crates/parser/src/grammar/items.rs

```diff
@@ -58,13 +58,25 @@ pub(super) fn item_or_macro(p: &mut Parser<'_>, stop_on_r_curly: bool) {
         Err(m) => m,
     };
 
-    if paths::is_use_path_start(p) {
+    // test macro_rules_as_macro_name
+    // macro_rules! {}
+    // macro_rules! {};
+    // macro_rules! ();
+    // macro_rules! [];
+    // fn main() {
+    //     let foo = macro_rules!();
+    // }
+
+    // test_err macro_rules_as_macro_name
+    // macro_rules! ()
+    // macro_rules! []
+    if paths::is_use_path_start(p)
+        || (p.at_contextual_kw(T![macro_rules]) && p.nth_at(1, BANG) && !p.nth_at(2, IDENT))
+    {
         match macro_call(p) {
-            BlockLike::Block => (),
-            BlockLike::NotBlock => {
-                p.expect(T![;]);
-            }
-        }
+            BlockLike::Block => p.eat(T![;]),
```

Comment:
```suggestion
            BlockLike::Block => (),
```
This is still wrong, `macro_call!{};` is only a valid item without the semicolon

Reply:
Ah, then I did change it correctly but need to adapt the test. Because I first had it without the eat, but then the test failed.